src/lib/queries/lexical-query.js

```diff
@@ -86,84 +106,72 @@ export default class LexicalQuery extends Query {
-      definitionRequests = definitionRequests.concat(requests.map(request => {
-        return {
-          request: request,
-          type: 'Short definition',
-          lexeme: lexeme,
-          appendFunction: 'appendShortDefs',
-          complete: false
-        }
-      }))
-      // Full definition requests
-      requests = this.lexicons.fetchFullDefs(lexeme.lemma, lexiconFullOpts)
-      definitionRequests = definitionRequests.concat(requests.map(request => {
-        return {
-          request: request,
-          type: 'Full definition',
-          lexeme: lexeme,
-          appendFunction: 'appendFullDefs',
-          complete: false
+    if (this.lemmaTranslations) {
+      let adapterTranslationRes = yield ClientAdapters.lemmatranslation.alpheios({
+        method: 'fetchTranslations',
+        params: {
+          homonym: this.homonym,
+          browserLang: this.lemmaTranslations.locale
         }
-      }))
-
-      lemmaList.push(lexeme.lemma)
-    }
+      })
+      if (adapterTranslationRes.errors.length > 0) {
+        adapterTranslationRes.errors.forEach(error => console.error(error))
+      }
 
-    if (this.lemmaTranslations) {
-      const languageCode = LMF.getLanguageCodeFromId(this.selector.languageID)
-      yield this.lemmaTranslations.adapter.fetchTranslations(lemmaList, languageCode, this.lemmaTranslations.locale)
       LexicalQuery.evt.LEMMA_TRANSL_READY.pub(this.homonym)
     }
 
     yield 'Retrieval of lemma translations completed'
 
-    // Handle definition responses
-    if (definitionRequests.length > 0) {
-      for (let definitionRequest of definitionRequests) {
-        definitionRequest.request.then(
-          definition => {
-            console.log(`${definitionRequest.type}(s) received:`, definition)
-            definitionRequest.lexeme.meaning[definitionRequest.appendFunction](definition)
-            definitionRequest.complete = true
-            if (this.active) {
-              LexicalQuery.evt.DEFS_READY.pub({
-                requestType: definitionRequest.type,
-                word: definitionRequest.lexeme.lemma.word,
-                homonym: this.homonym
-              })
-            }
-            if (definitionRequests.every(request => request.complete)) {
-              this.finalize('Success')
-            }
-          },
-          error => {
-            console.error(`${definitionRequest.type}(s) request failed: ${error}`)
-            definitionRequest.complete = true
-            if (this.active) {
-              LexicalQuery.evt.DEFS_NOT_FOUND.pub({
-                requestType: definitionRequest.type,
-                word: definitionRequest.lexeme.lemma.word
-              })
-            }
-            if (definitionRequests.every(request => request.complete)) {
-              this.finalize(error)
-            }
-          }
-        )
+    let adapterLexiconResShort = yield ClientAdapters.lexicon.alpheios({
```

Comment:
I think one thing we have lost here and below on line 149 is the ability to retrieve definitions from multiple lexicons concurrently (asynchronously) and populate the UI with the definitions *as they are retrieved* from each source, rather than waiting until all are retrieved. The code here is much easier to read and understand than it was before, but the asynchronous concurrency requirement for working with multiple lexicons was driving the prior more complex design.  I think implementing the suggestion @kirlat made for introducing an async interface to the Client Adapaters would help us put this back in in a cleaner way. I am fine with doing it in steps, but we need to not lose the the requirement to have that behavior supported here.

Reply:
Yes, you are right - I have lost this requirement.
I think if look at it closer - what is currently concurent here: it is precaching data for shortDefinitions, indexes for FullDefinitions and geting current data from FullDefinitions.
I think I could create concurrent queries inside ClientAdapter, make them parallell and send some events for updating definitions inside UIController.